beacon-chain/core/validators/validator.go

```diff
@@ -128,3 +128,71 @@ func SlashValidator(state *pb.BeaconState, slashedIdx uint64, whistleBlowerIdx u
 	state = helpers.IncreaseBalance(state, whistleBlowerIdx, whistleblowerReward-proposerReward)
 	return state, nil
 }
+
+// ActivatedValidatorIndices determines the indices activated during the current epoch.
+func ActivatedValidatorIndices(state *pb.BeaconState) []uint64 {
+	currentEpoch := helpers.CurrentEpoch(state)
+	activations := make([]uint64, 0)
+	delayedActivationEpoch := helpers.DelayedActivationExitEpoch(currentEpoch)
+	for i := 0; i < len(state.Validators); i++ {
+		val := state.Validators[i]
+		if val.ActivationEpoch == delayedActivationEpoch {
+			activations = append(activations, uint64(i))
+		}
+	}
+	return activations
+}
+
+// SlashedValidatorIndices determines the indices slashed during the current epoch.
+func SlashedValidatorIndices(state *pb.BeaconState) []uint64 {
+	currentEpoch := helpers.CurrentEpoch(state)
+	slashed := make([]uint64, 0)
+	for i := 0; i < len(state.Validators); i++ {
+		val := state.Validators[i]
+		maxWithdrawableEpoch := mathutil.Max(val.WithdrawableEpoch, currentEpoch+params.BeaconConfig().EpochsPerSlashingsVector)
+		if val.WithdrawableEpoch == maxWithdrawableEpoch && val.Slashed {
+			slashed = append(slashed, uint64(i))
+		}
+	}
+	return slashed
+}
+
+// ExitedValidatorIndices determines the indices exited during the current epoch.
+func ExitedValidatorIndices(state *pb.BeaconState) ([]uint64, error) {
```

Comment:
you loop through the validator set 3 times here, anyway we can do it all in a single iteration ?

Reply:
Will try to consolidate

Reply:
I can't find a way to break it up - it seems like each next iteration depends on the previous one to complete